orbit/pkg/setup_experience/setup_experience.go

```diff
@@ -186,19 +186,20 @@ func (s *SetupExperiencer) Run(oc *fleet.OrbitConfig) error {
 	if len(payload.Software) > 0 || payload.Script != nil {
 		log.Info().Msg("setup experience: rendering software and script UI")
 
-		var steps []*fleet.SetupExperienceStatusResult
 		if len(payload.Software) > 0 {
-			steps = payload.Software
-		}
-
-		if payload.Script != nil {
-			steps = append(steps, payload.Script)
+			for _, step := range payload.Software {
+				// If any step is not in a terminal state, then we're not done.
+				if (step.Status != fleet.SetupExperienceStatusFailure && step.Status != fleet.SetupExperienceStatusSuccess) ||
+					// If any software failed, and we're requiring all software to succeed, then we'll block completion.
+					(step.Status == fleet.SetupExperienceStatusFailure && payload.RequireAllSoftware) {
+					allStepsDone = false
+				}
+			}
 		}
 
-		for _, step := range steps {
-			if step.Status != fleet.SetupExperienceStatusFailure && step.Status != fleet.SetupExperienceStatusSuccess {
-				allStepsDone = false
-			}
+		// If a script is still running, then we're not done.
+		if payload.Script != nil && payload.Script.Status != fleet.SetupExperienceStatusFailure && payload.Script.Status != fleet.SetupExperienceStatusSuccess {
+			allStepsDone = false
 		}
```

Comment:
Because I'm not 100% sure what data will be sent with no-package software (do they have software title IDs?) I reworked this a bit so that it first checks everything returned in the `Software` array, and blocks setup experience if anything is failed while "RequireAllSoftware" is true.  Then it checks any `Script` data independently.

Reply:
@cdcme Can you weigh in?

Reply:
Thanks @cdcme! It works this way too and is a bit more explicit so I'll leave it as-is.